web/html/src/manager/virtualization/guests/console/guests-console.js

```diff
@@ -28,56 +27,93 @@ type State = {
   connected: boolean,
   popupState: string,
   password: ?string,
+  vmState: string,
+  currentToken: string,
+  hostId: string,
 };
 
 class GuestsConsole extends React.Component<Props, State> {
   client: ConsoleClientType;
-  inhibitPopup: boolean;
   popupSubmit: Function;
+  websocket: WebSocket;
+  pageUnloading: boolean;
+
+  clients = {
+    vnc: VncClient,
+    spice: SpiceClient,
+  };
 
   constructor(props: Props) {
     super(props);
 
-    const clients = {
-      vnc: VncClient,
-      spice: SpiceClient,
-    };
-    const port = window.location.port ? `:${window.location.port}` : "";
-    const url = `wss://${window.location.hostname}${port}/rhn/websockify/?token=${this.props.token}`;
-    this.client = new clients[this.props.graphicsType]('canvas', url, this.onConnect, this.onDisconnect, this.askPassword);
-
-    const editUrl = `/rhn/manager/systems/details/virtualization/guests/${this.props.hostId}/edit/${this.props.guestUuid}`;
-    const error = this.client !== undefined
+    const error = this.clients[props.graphicsType] != null
       ? undefined
-      : t(`Can not show display. Ensure, the virtual machine is stopped, <a href='${editUrl}'>set the display</a> to VNC and start again`);
+      : t(`Can not show display. Ensure, the virtual machine is stopped, set the display to VNC and start again`);
+
+    this.pageUnloading = false;
+    this.openVirtSocket();
 
     this.state = {
       error,
       expanded: false,
       connected: false,
       popupState: 'wait',
       password: undefined,
+      vmState: props.guestState,
+      currentToken: props.token,
+      hostId: props.hostId,
     };
   }
 
   componentDidMount() {
     this.connect();
 
     window.onbeforeunload = () => {
-      this.inhibitPopup = true;
+      this.pageUnloading = true;
+      if (this.client != null) {
+        this.client.removeErrorHandler();
+      }
```

Comment:
same as before, this could be rewritten to `this.client?.removeErrorHandler()`

Reply:
does this work here given it's not TS?

Reply:
This is weird: I believe it works in JS as well, but I don't find any specific documentation about it. Moreover it works on some sandbox and it doesn't on others. Better to not use it until it is officially supported. Sorry for the noise.

Reply:
Syntactic sugar like this is handled by Babel and is shared between JS and TS. For future reference you can check out `webpack.config.js` which has

```
test: /\.(ts|js)x?$/,
  exclude: /node_modules/,
  use: {
    loader: "babel-loader"
  }
}
```

(And `.babelrc` has the stuff like `"@babel/plugin-proposal-optional-chaining"` etc)

Reply:
OK, then I changed it. Tested here and it works